---
title: "Bioinformatics Class 14"
author: "Natalie Toothacre"
date: "May 17, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Transcriptomics and the analysis of RNA-Seq data

See: https://bioboot.github.io/bimm143_S18/class-material/lecture14-BIMM143_W18-lab/

Input countData and coData into R

```{r}
counts <- read.csv("data/airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <- read.csv("data/airway_metadata.csv", stringsAsFactors = FALSE)
```

```{r}
head(counts)
```

```{r}
head(metadata)
```

# Toy differential gene expression

```{r}
View(metadata)
```


Calculate mean counts per gene across control samples

```{r}
control <- metadata[metadata[,"dex"]=="control",]
control.mean <- rowSums( counts[ ,control$id] )/4 
names(control.mean) <- counts$ensgene
```

Calculate mean counts per gene across treated samples

```{r}
treated <- metadata[metadata[,"dex"]=="treated",]
treated.mean <- rowSums( counts[ ,treated$id] )/4 
names(treated.mean) <- counts$ensgene
```

Combine meancounts data

```{r}
meancounts <- data.frame(control.mean, treated.mean)
head(meancounts)
```

Make a scatterplot

```{r}
plot.default(control.mean, treated.mean, log = "xy")
```


We can find candidate differentially expressed genes by looking for genes with a large change between control and dex-treated samples. We usually look at the log2 of the fold change, because this has better mathematical properties.

Here we calculate log2foldchange, add it to our meancounts data.frame

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```


The NaN is returned when you divide by zero and try to take the log. The -Inf is returned when you try to take the log of zero. It turns out that there are a lot of genes with zero expression. Let’s filter our data to remove these genes.


```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```


Differential expression is generally above 2 or below -2 thresholds

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```


```{r}
paste("the number of upregulated genes is", sum(up.ind))
paste("the number of downregulated genes is", sum(down.ind))
```

# Adding annotation data

```{r}
anno <- read.csv("data/annotables_grch38.csv")
head(anno)
```

> **Q6.** From consulting the help page for the merge() function can you set the by.x and by.y arguments appropriately to annotate our mycounts data.frame with all the available annotation data in your anno data.frame?




```{r}
results <- merge(mycounts, anno, by.x = "row.names", by.y = "ensgene")
head(results)
```


```{r}
#biocLite("AnnotationDbi")
#biocLite("org.Hs.eg.db")
library("AnnotationDbi")
library("org.Hs.eg.db")
```


```{r}
columns(org.Hs.eg.db)
```




#DESeq 2 Analysis

```{r}
#biocLite("DESeq2")

library(DESeq2)
citation("DESeq2")
```


```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```



```{r}
dds <- DESeq(dds)
```


```{r}
res <- results(dds)
res
```


```{r}
summary(res)
```


```{r}
resOrdered <- res[order(res$pvalue),]
```


```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```

The more generic way to access the actual subset of the data.frame passing a threshold like this is with the subset() function

```{r}
resSig05 <- subset(as.data.frame(res), padj < 0.05)
nrow(resSig05)
```

```{r}
resSig01 <- subset(as.data.frame(res), padj <0.01)
nrow(resSig01)
```

```{r}
head(resSig01)
```

> **Q10.** Using either the previously generated anno object (annotations from the file annotables_grch38.csv file) or the mapIds() function (from the AnnotationDbi package) add annotation to your res01 results data.frame.



```{r}
resSig01$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(resSig01),
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
```


You can arrange and view the results by the adjusted p-value

```{r}
ord <- order( resSig01$padj )
#View(res01[ord,])
head(resSig01[ord,])
```

```{r}
write.csv(resSig01[ord,], "signif01_results.csv")
```


# Data Visualization


Let’s first see what the gene ID is for the CRISPLD2 gene using

```{r}
i <- grep("CRISPLD2", resSig01$symbol)
resSig01[i,]
```


```{r}
rownames(resSig01[i,])
```

Now, with that gene ID in hand let’s plot the counts, where our intgroup, or “interesting group” variable is the “dex” column.

```{r}
plotCounts(dds, gene="ENSG00000103196", intgroup="dex")
```


Notice that we could have actually returned the data instead of plotting. We could then pipe this to ggplot and make our own figure. Let’s make a boxplot.


```{r}
d <- plotCounts(dds, gene="ENSG00000103196", intgroup="dex", returnData=TRUE)
head(d)
```

```{r}
boxplot(count ~ dex , data=d)
```













